src/lib/proof-system/zkFunction.ts

```diff
@@ -0,0 +1,157 @@
+import { Snarky, initializeBindings } from '../../snarky.js';
+import { MlFieldArray, MlFieldConstArray } from '../ml/fields.js';
+import { withThreadPool } from '../../snarky.js';
+import { Provable } from '../provable/provable.js';
+import { snarkContext, gatesFromJson } from '../provable/core/provable-context.js';
+import { prettifyStacktrace, prettifyStacktracePromise } from '../util/errors.js';
+import { ProvablePure } from '../provable/types/provable-intf.js';
+
+// external API
+export { ZkFunction, Keypair, Proof, VerificationKey };
+
+type ZkFunctionConfig<P, W extends any[]> = {
+  name: string;
+  publicInputType: ProvablePure<P>;
+  privateInputTypes: { [K in keyof W]: ProvablePure<W[K]> };
+  main: (publicInput: P, ...privateInputs: W) => void;
```

Comment:
In ZkProgram, the layout of the `main` function (or methods) depends on the type of `publicInputType` - we can derive the shape of the method based on whether we have a provable type as part of the ZkFunction config and type or simply leave the property "empty" (undefined) 

```ts
type Method<PublicInput> = PublicInput extends undefined
  ? {
      method(...args: any): Promise<any>;
    }
  : {
      method(publicInput: PublicInput, ...args: any): Promise<any>;
    };

type MyMethod1 = Method<undefined>;
        // ^ type MyMethod1 = { method(...args: any): Promise<any>; }
type MyMethod2 = Method<Field>;
        // ^ type MyMethod2 = { method(publicInput: Field, ...args: any): Promise<any>; }
```